generic-sensor/generic-sensor-tests.js

```diff
@@ -36,84 +36,81 @@ function reading_to_array(sensor) {
+
+    await eventWatcher.wait_for("reading");
+    assert_reading_not_null(sensor);
+    assert_true(sensor.hasReading);
+
+    sensor.stop();
+    assert_reading_null(sensor);
+    assert_false(sensor.hasReading);
   }, `${sensorType.name}: Test that 'onreading' is called and sensor reading is valid`);
 
-  async_test(t => {
-    let sensor1 = new sensorType();
-    let sensor2 = new sensorType();
-    sensor1.onreading = t.step_func_done(() => {
-      // Reading values are correct for both sensors.
-      assert_reading_not_null(sensor1);
-      assert_reading_not_null(sensor2);
-
-      //After first sensor stops its reading values are null,
-      //reading values for the second sensor remains
-      sensor1.stop();
-      assert_reading_null(sensor1);
-      assert_reading_not_null(sensor2);
-      sensor2.stop();
-      assert_reading_null(sensor2);
-    });
-    sensor1.onerror = t.step_func_done(unreached);
-    sensor2.onerror = t.step_func_done(unreached);
+  promise_test(async t => {
+    const sensor1 = new sensorType();
+    const sensor2 = new sensorType();
+    const eventWatcher = new EventWatcher(t, sensor1, ["reading", "error"]);
     sensor2.start();
     sensor1.start();
+
+    await eventWatcher.wait_for("reading");
+    // Reading values are correct for both sensors.
+    assert_reading_not_null(sensor1);
+    assert_reading_not_null(sensor2);
+
+    //After first sensor stops its reading values are null,
+    //reading values for the second sensor remains
+    sensor1.stop();
+    assert_reading_null(sensor1);
+    assert_reading_not_null(sensor2);
+    sensor2.stop();
+    assert_reading_null(sensor2);
   }, `${sensorType.name}: sensor reading is correct`);
 
-  async_test(t => {
-    let sensor = new sensorType();
-    let cachedTimeStamp1;
-    sensor.onreading = () => {
-      cachedTimeStamp1 = sensor.timestamp;
-    };
-    sensor.onerror = t.step_func_done(unreached);
+  promise_test(async t => {
+    const sensor = new sensorType();
+    const eventWatcher = new EventWatcher(t, sensor, ["reading", "error"]);
     sensor.start();
-    t.step_timeout(() => {
-      sensor.onreading = t.step_func_done(() => {
-        //sensor.timestamp changes.
-        let cachedTimeStamp2 = sensor.timestamp;
-        assert_greater_than(cachedTimeStamp2, cachedTimeStamp1);
-        sensor.stop();
-      });
-    }, 1000);
+
+    await eventWatcher.wait_for("reading");
+    const cachedTimeStamp1 = sensor.timestamp;
+
+    await eventWatcher.wait_for("reading");
+    const cachedTimeStamp2 = sensor.timestamp;
+
+    assert_greater_than(cachedTimeStamp2, cachedTimeStamp1);
+    sensor.stop();
   }, `${sensorType.name}: sensor timestamp is updated when time passes`);
 
-  async_test(t => {
-    let sensor = new sensorType();
-    sensor.onerror = t.step_func_done(unreached);
+  promise_test(async t => {
+    const sensor = new sensorType();
+    const eventWatcher = new EventWatcher(t, sensor, ["reading", "error"]);
```

Comment:
Let's use 'activate' event to monitor sensor state.

Reply:
You are right, update to 'activate'.